### PR TITLE
Generalize errors so they are consumed in all authenticators

### PR DESCRIPTION
**Pull request by orenbm**
This PR re-arranges some errors so they can be consumed easily in all authenticators.

This PR also changes the behaviour of the Kubernetes authenticator so it now returns 403 on unpermitted hosts instead of a 401. This was done to align it with the other authenticators.

### CODE REVIEW

File: app/domain/authentication/authn_k8s/validate_pod_request.rb [hunk 1]
```diff
@@ -73,7 +73,7 @@ def host_can_access_service?
 
       def host
         @host ||= @resource_class[k8s_host.conjur_host_id]
-        raise Err::HostNotFound(k8s_host.conjur_host_id) if @host.nil?
+        raise ::Errors::Authentication::HostNotFound(k8s_host.conjur_host_id) if @host.nil?
```
[the remark this answers is not in the document]
Comment (sigalsax) [reply app/domain/authentication/authn_k8s/validate_pod_request.rb]: ^ I am not sure what that means or why that might be an issue. do you?

**Comment by sigalsax** [comment]
Good idea! If we already here, let's think about errors that currently exist in AuthnK8s and if they can be applicable to AuthnAzure. 

For example:
"Err::HostNotAuthorized" to check if the resource can authenticate to Conjur? 
"WebserviceNotFound" to check if the authenticator exists in Conjur as a webservice. Aka if the Azure authn was loaded as a webservice in Conjur

File: app/domain/authentication/authn_k8s/validate_pod_request.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (orenbm) [reply app/domain/authentication/authn_k8s/validate_pod_request.rb]: you can ignore this comment

File: app/domain/errors.rb [hunk 2]
```diff
@@ -73,24 +73,24 @@ module AuthenticatorClass
 
       module Security
 
-        NotWhitelisted = ::Util::TrackableErrorClass.new(
+        AuthenticatorNotWhitelisted = ::Util::TrackableErrorClass.new(
           msg: "'{0-authenticator-name}' is not whitelisted in CONJUR_AUTHENTICATORS",
```
Comment (sigalsax) [inline app/domain/errors.rb]: Because the whitelisting happens in DB, this message ("whitelisted in CONJUR_AUTHENTICATORS") I believe is not correct. How does enabling look in the DB?

File: app/domain/errors.rb [hunk 3]
```diff
@@ -73,24 +73,24 @@ module AuthenticatorClass
...
-          code: "CONJ00006E"
+        RoleNotFound = ::Util::TrackableErrorClass.new(
+          msg: "'{0-role-name}' wasn't found",
+          code: "CONJ00007E"
         )
 
-        UserNotDefinedInConjur = ::Util::TrackableErrorClass.new(
-          msg: "User '{0-user-name}' is not defined in Conjur",
-          code: "CONJ00007E"
+        RoleNotAuthorizedOnWebservice = ::Util::TrackableErrorClass.new(
```
Comment (sigalsax) [inline app/domain/errors.rb]: please move this above "RoleNotFound" so it is in the proper numerical order

File: app/controllers/concerns/basic_authenticator.rb [hunk 4]
```diff
@@ -19,7 +19,7 @@ def perform_basic_authn
       raise ApplicationController::Unauthorized, "Invalid username or password"
     rescue Errors::Authentication::InvalidOrigin
       raise ApplicationController::Forbidden, "User is not authorized to login from the current origin"
-    rescue Errors::Authentication::Security::UserNotAuthorizedInConjur
```
Comment (sigalsax) [inline app/controllers/concerns/basic_authenticator.rb]: Please follow suite here by setting  "Errors::Authentication::Security" outside so that it is "Err = Errors::Authentication::Security"

Comment (sigalsax) [reply app/controllers/concerns/basic_authenticator.rb]: *below*

File: cucumber/kubernetes/features/login_errors.feature [hunk 5]
```diff
@@ -15,12 +15,12 @@ Feature: Errors emitted by the login method.
...
     on the webservice.
     When I login to authn-k8s as "pod/inventory-unauthorized"
-    Then the HTTP status is "401"
+    Then the HTTP status is "403"
 
   Scenario: it raises an error when logging in with a custom prefix as a host which does not
     hold "authenticate" privilege on the webservice.
     When I login to authn-k8s as "@namespace@/pod/inventory-unauthorized" with prefix "host/conjur/authn-k8s/minikube/apps"
-    Then the HTTP status is "401"
+    Then the HTTP status is "403"
```
Comment (sigalsax) [inline cucumber/kubernetes/features/login_errors.feature]: how did our tests pass previously if the error status wasn't correct?

File: spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb [hunk 6]
```diff
@@ -87,15 +87,15 @@
     subject(:validator) { Authentication::AuthnK8s::ValidatePodRequest
                             .new(dependencies: dependencies) }
 
-    it 'raises WebserviceNotFound error when webservice is missing' do
+    it 'raises WebserviceNotDefined error when webservice is missing' do
```
Comment (sigalsax) [inline spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb]: Why was this renamed from WebserviceNotFound -> WebserviceNotDefined here? Did a search and "WebserviceNotDefined" doesn't exist in other places but this one

File: spec/app/domain/authentication/validate_webservice_access_spec.rb [hunk 7]
```diff
@@ -92,7 +92,7 @@ def role_class(returned_role)
       )
     end
     it "raises a NotDefinedInConjur error" do
```
Comment (sigalsax) [inline spec/app/domain/authentication/validate_webservice_access_spec.rb]: This needs to be changed here b/c we changed it to "Role not found"

File: spec/app/domain/authentication/validate_webservice_access_spec.rb [hunk 8]
```diff
@@ -112,7 +112,7 @@ def role_class(returned_role)
     end
 
     it "raises a NotAuthorizedInConjur error" do
```
Comment (sigalsax) [inline spec/app/domain/authentication/validate_webservice_access_spec.rb]: same here. We need to change the expectation here b/c we changed it to RoleNotAuthorizedOnA*Webservice*

File: app/domain/errors.rb (hunk 2 above)
Comment (orenbm) [reply app/domain/errors.rb]: changed to `'{0-authenticator-name}' is not enabled`

File: app/controllers/concerns/basic_authenticator.rb (hunk 4 above)
Comment (orenbm) [reply app/controllers/concerns/basic_authenticator.rb]: This class doesn't use this method. We do this only when we have a lot of the same error module in the class so we don't want to repeat ourselves.

File: cucumber/kubernetes/features/login_errors.feature (hunk 5 above)
Comment (orenbm) [reply cucumber/kubernetes/features/login_errors.feature]: This PR actually changes the returned error to a better one. 403 Forbidden is a proper error in this use-case.

File: spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb (hunk 6 above)
Comment (orenbm) [reply spec/app/domain/authentication/authn_k8s/validate_pod_request_spec.rb]: that's a mistake, it should be `WebserviceNotFound`

File: app/domain/errors.rb [hunk 9]
```diff
@@ -73,23 +73,23 @@ module AuthenticatorClass
...
-        UserNotAuthorizedInConjur = ::Util::TrackableErrorClass.new(
-          msg: "User '{0-user-name}' is not authorized in the Conjur policy",
+        RoleNotAuthorizedOnWebservice = ::Util::TrackableErrorClass.new(
+          msg: "'{0-role-name}' does not have 'authenticate' privilege on {1-service-name}",
           code: "CONJ00006E"
         )
 
-        UserNotDefinedInConjur = ::Util::TrackableErrorClass.new(
-          msg: "User '{0-user-name}' is not defined in Conjur",
+        RoleNotFound = ::Util::TrackableErrorClass.new(
```
Comment (sigalsax) [inline app/domain/errors.rb]: Maybe we should leave "In Conjur" to be clear

Comment (orenbm) [reply app/domain/errors.rb]: I removed it because we don't want the word `Conjur` in our logs anyway
